### DISCUSSION TITLE
Adapting the ARI format for relative rewriting & complexity

### DISCUSSION THREAD

**Discussion by ffrohn** [TermCOMP/starexec-master#85] [General]
Since we want to [use the ARI format](https://github.com/orgs/TermCOMP/discussions/83) in the future, we have to extend it for relative rewriting. The idea is to use the format for [multiple TRSs](http://project-coco.uibk.ac.at/ARI/multiple-trs.php), with the restriction that there have to be precisely two TRSs. Then the goal is to prove termination of TRS 1 relative to TRS 2.

What do you think?

Edit (24/01/16): Taking the discussion into account, the following proposal seems to be better:
```
RTRS ::= ( format RTRS ) fun* rule*
 fun ::= ( fun identifier number )
rule ::= ( rule term term cost? )
cost ::= ( :cost number )
```
Note that `number` only allows for non-negative numbers in [ARI](http://project-coco.uibk.ac.at/ARI/index.php). For termination, it only matters whether the costs are zero or not. For complexity, the costs only matter for concrete (non-asymptotic) bounds.

I'm not sure if RTRS (relative TRS) is a good name, though. We could also go for, e.g., WTRS (weighted TRS).

**Comment by stefandollase** [comment]
I think this is a good suggestion when only considering the cost 0 and 1. However, I would be interested in having a format that uses an arithmetic term as the cost (and also a condition). In this way, relative TRSs, ITSs and ITRSs can use the same format. Of course one could do it in a separate format, but then again one would run into the situation that was already described [here](https://github.com/orgs/TermCOMP/discussions/87#discussioncomment-8141185), i.e., if the ITRS is incidentally a relative TRS, then one has to encode the ITRS in a specialized relative rewriting format, so it can be understood by tools that can only handle relative rewriting but no ITRSs.

**Reply by ffrohn** [reply]
But if it's a (relative) TRS, then the cost-term can only be a constant `c` (as there are no variables of type `int`). Then for termination, it only matters whether `c` is positive or zero (assuming that we only talk about non-negative costs). For complexity, the actual value of `c` only matters if you're interested in concrete (non-asymptotic) bounds. But there's hardly tool support for concrete bounds, is it?

**Reply by stefandollase** [reply]
Indeed, I am not interested in concrete bounds, but in a format for relative rewriting that is consistent with a format for ITRSs. In ITRSs, there are int-variables, which can occur in the cost term, so it's not a constant. It is correct, that in relative rewriting the cost will always be a constant. In that way relative rewriting is a special case of ITRSs. My suggestion is, to use a restricted ITRS syntax for relative rewriting instead of a dedicated syntax.

**Reply by AkihisaYamada** [reply]
I'm very interested in cost-annotated rewriting, actually... A potential suggestion is to allow `(rule l r :cost 0)`, having extension to non-zero and non-constant costs in mind.
However, this is only one particular view. I feel viewing RTRS just as a pair of TRSs is also very reasonable, and not needing to modify ARI-infra is a big merit.

**Reply by ffrohn** [reply]
If several people are interested in cost-annotations, why not? But now we're only talking about the complexity categories, not about termination of relative rewriting, right? And we're only talking about non-negative costs, correct?

**Reply by AkihisaYamada** [reply]
Ah, I was speaking about termination. Do relative rules appear already in complexity categories? Then I think cost-annotation is clearer for them. And I'd use the same format both for termination and complexity.

**Reply by ffrohn** [reply]
Yes, relative rules are allowed in the complexity categories. But what's the purpose of cost-annotations for termination?

**Reply by stefandollase** [reply]
Indeed, I am only interested in non-negative costs.

>But now we're only talking about the complexity categories, not about termination of relative rewriting, right?

I don't understand the question. Shouldn't the language to describe a rewrite system be independent of the analysis goal, especially if we separate these two parts to reuse rewrite systems with different analysis goals? One could also analyze an ITRS with non-constant cost terms for termination, where not only a normal form, but also an infinite sequence with only a finite number of non-0-cost steps is considered terminating.

**Reply by ffrohn** [reply]
Yes, we should have a uniform format. I'm just trying to understand where the *motivation* for cost-annotations is coming from.

**Comment by AkihisaYamada** [comment]
> But what's the purpose of cost-annotations for termination?

No, for termination, only `:cost 0` make difference.

**Reply by stefandollase** [reply]
Couldn't the `0` also be the result of evaluating a cost-term?

**Comment by ffrohn** [comment]
I think we are discussing two different things here:

Stefan is mainly interested in complexity of ITRSs with non-constant cost-expressions.

Akihisa is interested in relative TRSs with constant cost-annotations.

@stefandollase I understand your motivation -- you want to have a format for relative TRSs which is a special case of ITRSs. However, there are two problems that need to be resolved first, in my opinion:
- there is no proposal for an ARI format for ITRSs
- there is no complexity category for ITRSs

@AkihisaYamada I don't fully understand your motivation yet, as you said that
1. you're "very interested in cost-annotated rewriting",
2. you're "speaking about termination", and
3. "for termination, only :cost 0 make difference."

For me, the last statement implies that cost-annotations are not very relevant for termination, but the first two seem to imply that you are very interested in termination of cost-annotated rewriting. What am I getting wrong?

**Reply by AkihisaYamada** [reply]
You are right. My true motivation for cost-annotation is (probabilistic) cost analysis, but currently only concerned with relative TRSs.

**Reply by ffrohn** [reply]
Thanks for the clarification!

**Reply by stefandollase** [reply]
>However, there are two problems that need to be resolved first, in my opinion:
>
>*    there is no proposal for an ARI format for ITRSs
>*    there is no complexity category for ITRSs

Indeed, these are still open issues. However, I want to ensure that in the future it will be possible to define an ARI ITRS format that is a generalization of the relative TRS format, i.e., that allows for cost terms. Having two separate TRSs for relative TRSs would not be helpful for that.

**Reply by stefandollase** [reply]
Would it be important to specify the ITRS format now?

**Reply by ffrohn** [reply]
I don't think so, it has been a demo category for a while and, as far as I'm aware, nobody is actively working on termination tools for ITRSs. So I think we can postpone that until there are new tools / techniques.

**Reply by cfuhs** [reply]
ITRSs are de facto a special case of LCTRSs, for which [ARI has a dedicated format](http://project-coco.uibk.ac.at/ARI/lctrs.php). To avoid a proliferation of different formats, it could be good to use this LCTRS format also for ITRSs. Termination analysis of LCTRSs is supported, e.g., by the tool [Ctrl](http://cl-informatik.uibk.ac.at/software/ctrl/), and there is active work on the tools [Cora](https://github.com/hezzel/cora/) and [Crisys2](https://git.trs.css.i.nagoya-u.ac.jp/nishida/crisys2/).

**Reply by ffrohn** [reply]
Let's move this to a [separate discussion](https://github.com/orgs/TermCOMP/discussions/89).

**Comment by ffrohn** [comment]
I updated the proposal to take the discussion into account.

**Reply by stefandollase** [reply]
Thanks! I like the new suggestion. In the future for ITRSs, we could then allow cost-terms instead of just constant numbers for the cost.

**Comment by ffrohn** [comment]
Everybody seems to be happy with the updated proposal --> closed.
